### PR TITLE
[WIP] Add client prototype

### PR DESCRIPTION
**Pull request by drets**

### CODE REVIEW

File: src/Network/MQTT/Client.hs [hunk 1]
```diff
@@ -0,0 +1,190 @@
+{-# LANGUAGE RecordWildCards #-}
+module Network.MQTT.Client
+    ( conn
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Shouldn't you export more functions?

File: src/Network/MQTT/Client.hs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+    hSetBuffering h (BlockBuffering Nothing)
+    hSetBinaryMode h True
+    return (sock, h)
+
+sendPacket :: SocketHandle -> Packet -> IO ()
+sendPacket socketHandle packet = do
+    hPutBuilder handle (encodePacket packet)
+    hFlush handle
+  where
+    handle = snd socketHandle
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: You can pattern-match handle in the argument list

File: src/Network/MQTT/Client.hs [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+    threadDelay sec
+    sender clientState publishPacketQoS1
+
+  threadDelay $ 3 * sec
+
+------------------------------------------------------------------------------------------------
+
+sec = 1000000
+
+data ClientState =
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Put data types on top of the file.

File: src/Network/MQTT/Client.hs [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+  ClientState
+  { senderQueue    :: TQueue Packet
+  , receiverQueue  :: TQueue Packet
+  , qosLevel       :: TVar (Maybe QoS)
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: What that QoS level means?

File: src/Network/MQTT/Client.hs [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+  , sessionPresent :: TVar (Maybe Bool)
+  , pingResp       :: TVar (Maybe Bool)
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Could you please write some haddock on how all these fields are used?

File: src/Network/MQTT/Client.hs [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+closeConn sh = print "Connection closed" >> hClose (snd sh)
+
+------------------------------------------------------------------------------------------------
+
+receiverSTM :: ClientState -> STM ()
+receiverSTM ClientState{..} = do
+  p <- readTQueue receiverQueue
+  case p of
+    (CONNACK ConnackPacket{..}) -> do
+      assert (connackReturnCode == Accepted)
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: You must close connection if `connackReturnCode /= Accepted`

File: src/Network/MQTT/Client.hs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        ("Connection is not accepted by Server. Status code: " ++ show connackReturnCode)
+      writeTVar sessionPresent (Just connackSessionPresent)
+    (PINGRESP _) -> do
+      sessionPresentFlag <- readTVar sessionPresent
+      assert (isJust sessionPresentFlag)
+        "Connack should be received first"
+      writeTVar pingResp (Just True)
+    (PUBACK PubackPacket{..}) -> do
+      qosLevelVal <- readTVar qosLevel
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Check `sessionPresentFlag`

File: src/Network/MQTT/Client.hs (hunk 4 above)
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Well, I see. You shouldn't just store single last QoS server required. Instead, you should use packet identifier to match requests with responses.

File: src/Network/MQTT/Client.hs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+  atomically $ receiverSTM state
+  receiver state
+
+sender :: ClientState -> Packet -> IO ()
+sender ClientState{..} p = do
+  case p of
+    (PINGREQ _)                 -> atomically $ writeTVar pingResp (Just False)
+    (PUBLISH PublishPacket{..}) -> do
+      sessionPresentFlag <- atomically $ readTVar sessionPresent
+      assert (isJust sessionPresentFlag) "Connack should be received first"
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: You shouldn't wait to receive connack before sending other packets.

File: src/Network/MQTT/Client.hs [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+  PUBLISH PublishPacket { publishDup = True
+                        , publishMessage = Message QoS1 True (Topic $ T.pack "a/b") (BS.pack [ 0xab, 0x12 ])
+                        , publishPacketIdentifier = Just (PacketIdentifier 10)
+                        }
+
+
+publishPacketQoS2 =
+  PUBLISH PublishPacket { publishDup = True
+                        , publishMessage = Message QoS2 True (Topic $ T.pack "a/b") (BS.pack [ 0xab, 0x12 ])
+                        , publishPacketIdentifier = Just (PacketIdentifier 10)
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Packets should have distinct packet ids. Otherwise, they may clash.

File: src/Network/MQTT/Client.hs [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+                        , connectWillMsg = Just $
+                            Message QoS1
+                                    False
+                                    (Topic $ T.pack "a/b")
+                                    (BS.pack [ 0x12, 0xab ])
+                        , connectUserName = Just (UserName $ T.pack "ac")
+                        , connectPassword = Just (Password $
+                                                      BS.pack [ 0xcd, 0xbb, 0x11 ])
+                        , connectCleanSession = True
+                        , connectKeepAlive = 0x000c
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: You don't send pingreq periodically to keep session active, so keepalive should be set to 0 (disabled).

File: src/Network/MQTT/Client.hs [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+  ClientState
+  { -- | All packets from the this queue will be sent to the Server
+    sendQueue    :: TQueue Packet
+    -- | Whether the Client is connected to the Server
+  , isConnected  :: TVar Bool
+  }
+  deriving (Eq)
+
+initState :: STM ClientState
+initState = do
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]:
```
initState :: STM ClientState
initState = ClientState <$> newTQueue <*> newTVar False
```

File: src/Network/MQTT/Client.hs [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+  sendQueue <- newTQueue
+  isConn <- newTVar False
+  return $ ClientState
+             sendQueue
+             isConn
+
+conn :: HostName -> String -> IO SocketHandle
+conn hostname port = do
+    addrinfos <- getAddrInfo Nothing (Just hostname) (Just port)
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]:
```
serveraddr : _ <- getAddrInfo Nothing (Just hostname) (Just port)
```

File: src/Network/MQTT/Client.hs [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+  sender' state socketHandle
+
+receiver' :: ClientState -> SocketHandle -> IO ()
+receiver' state@ClientState{..} socketHandle = do
+  (_, pkgs) <- parseMany (fst socketHandle) (parse parsePacket) (parse parsePacket)
+  forM_ pkgs $ \p -> do
+    print $ "receive: " ++ show p
+    case p of
+      (CONNACK ConnackPacket{..}) -> do
+        when (connackReturnCode /= Accepted) (disconn socketHandle)
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: You shouldn't send `DISCONNECT`, just close the socket.

File: src/Network/MQTT/Client.hs [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+  print $ "send: " ++ show packet
+  sender' state socketHandle
+
+receiver' :: ClientState -> SocketHandle -> IO ()
+receiver' state@ClientState{..} socketHandle = do
+  (_, pkgs) <- parseMany (fst socketHandle) (parse parsePacket) (parse parsePacket)
+  forM_ pkgs $ \p -> do
+    print $ "receive: " ++ show p
+    case p of
+      (CONNACK ConnackPacket{..}) -> do
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Parentheses are not needed

File: src/Network/MQTT/Client.hs (hunk 11 above)
Comment (drets) [inline src/Network/MQTT/Client.hs]: 👍

File: src/Network/MQTT/Client.hs [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+        when (connackReturnCode /= Accepted) (disconn socketHandle)
+        atomically $ writeTVar connected True
+      (PINGRESP PingrespPacket) -> atomically $ writeTVar pingRespReceived True
+      _ -> return ()
+  receiver' state socketHandle
+
+sender :: ClientState -> Packet -> IO ()
+sender ClientState{..} p = atomically $ do
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: I would create `sendPingreq` method, as that will be really hard to extend this one to other types of packets.

File: src/Network/MQTT/Client.hs (hunk 12 above)
Comment (drets) [inline src/Network/MQTT/Client.hs]: 👍

File: src/Network/MQTT/Client.hs (hunk 15 above)
Comment (drets) [inline src/Network/MQTT/Client.hs]: Like [this](https://github.com/rasendubi/mqtt-broker/pull/25/commits/117ededd9812783368d3dd45e1f582eaf45a6760#diff-2d699f0b2ca885a6b29074801ffcacf6R98)?

Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: yep

File: src/Network/MQTT/Client.hs [hunk 16]
```diff
@@ -18,31 +20,40 @@ import           Data.ByteString.Builder       ( hPutBuilder )
...
-  , connected        :: TVar Bool -- ^ Whether the Client is connected to the Server
-  , sessionPresent   :: TVar Bool -- ^ Whether the Server has Session for the Client
-  , pingRespReceived :: TVar Bool -- ^ Whether the Client received Pingresp packet from the Server
-  }
-  deriving (Eq)
+  { sendQueue                 :: TQueue Packet -- ^ All packets from the this queue will be sent to the Server
+  , connected                 :: TVar Bool -- ^ Whether the Client is connected to the Server
+  , sessionPresent            :: TVar Bool -- ^ Whether the Server has Session for the Client
+  , pingRespReceived          :: TVar Bool -- ^ Whether the Client received Pingresp packet from the Server
+  , notAcknSubscribePackets   :: TVar [SubscribePacket] -- ^ Keep Subscribe packets which have not been acknowledged
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: why `Ackn`?

File: src/Network/MQTT/Client.hs [hunk 17]
```diff
@@ -90,6 +101,32 @@ receiver' state@ClientState{..} socketHandle = do
                   writeTVar connected True
                   writeTVar sessionPresent connackSessionPresent
       PINGRESP PingrespPacket -> atomically $ writeTVar pingRespReceived True
+      SUBACK SubackPacket{..} -> atomically $ do
+        -- | Get current Subscribe packets which were not acknowledged
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Haddock comments won't work on statements.

File: src/Network/MQTT/Client.hs (hunk 16 above)
Comment (drets) [inline src/Network/MQTT/Client.hs]: How do you propose? :)

File: src/Network/MQTT/Client.hs [hunk 18]
```diff
@@ -90,6 +101,32 @@ receiver' state@ClientState{..} socketHandle = do
...
+        writeTVar notAcknSubscribePackets notAcknSubscribePackets'
+        -- | Get topic filters from the Subscribe packet which was acknowledged
+        let topicFilters = fst <$> subscribeTopicFiltersQoS acknSubscribePacket
+        -- | Generate topic filter and Subscription pairs.
+        -- 1. Zip topic filters from the Subscribe packet which was acknowledged
+        -- with Suback responses.
+        -- 2. Filter Failure Suback responses
+        -- 3. Create Subscription out of QoS
+        let topicFiltersSubscription =
+              (\(topicFilter, qos) -> (topicFilter, Subscription (fromJust qos))) <$> filter
```
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: Consider using `fmap (\(topicFilter, qos) -> ...) . filter (\(_, subackResponse) -> ...) $ zip topicFilters subackResponses`

File: src/Network/MQTT/Client.hs (hunk 16 above)
Comment (rasendubi) [inline src/Network/MQTT/Client.hs]: `notAckSubscribePackets`. I meant `Acknowledgement` is commonly shortened as `Ack`, not `Ackn`.

Comment (drets) [inline src/Network/MQTT/Client.hs]: 👍

File: test/Network/MQTT/ClientSpec.hs [hunk 19]
```diff
@@ -1,37 +1,69 @@
...
       p <- S.read os'
       p `shouldBe` Just (PINGREQ PingreqPacket)
+
+    it "should call user callback on Pingresp" $ do
+      (is, is') <- S.makeChanPipe
+      (os', os) <- S.makeChanPipe
+      client <- runClient is os
+      p <- newEmptyMVar
+      setPingCallback (state client) (Just $ putMVar p True)
+      S.write (Just $ PINGRESP PingrespPacket) is'
```
Comment (rasendubi) [inline test/Network/MQTT/ClientSpec.hs]: The client haven't sent Pingreq

File: test/Network/MQTT/ClientSpec.hs [hunk 20]
```diff
@@ -1,37 +1,69 @@
...
+      res <- takeMVar p
+      res `shouldBe` True
+
+    it "should call user callback on Publish" $ do
+      debugTests
+      (is, is') <- S.makeChanPipe
+      (os', os) <- S.makeChanPipe
+      client <- runClient is os
```
Comment (rasendubi) [inline test/Network/MQTT/ClientSpec.hs]: The next pattern is worth abstracting
```haskell
(is, is') <- S.makeChanPipe
(os', os) <- S.makeChanPipe
client <- runClient is
```

File: test/Network/MQTT/ClientSpec.hs [hunk 21]
```diff
@@ -1,37 +1,69 @@
...
+
+      let expectedTopicName = Topic $ T.pack "a/b"
+      let expectedMessage = BS.pack [0xab, 0x12]
+      actualTopicName' <- newEmptyMVar
+      actualMessage' <- newEmptyMVar
+      setPublishCallback (state client) $ Just $ \t m -> do
+        putMVar actualTopicName' t
+        putMVar actualMessage' m
+      flip S.write is' $ Just $ PUBLISH
```
Comment (rasendubi) [inline test/Network/MQTT/ClientSpec.hs]: The client hasn't subscribed, so the flow may be invalid.

File: test/Network/MQTT/ClientSpec.hs (hunk 19 above)
Comment (drets) [inline test/Network/MQTT/ClientSpec.hs]: Should I close connection in this case?

File: src/Network/MQTT/Client/Core.hs [hunk 22]
```diff
@@ -7,45 +7,53 @@ module Network.MQTT.Client.Core
...
 import qualified Data.Map.Strict as Map
 import           Data.Maybe ( isJust, fromJust )
 import qualified Data.Sequence as Seq
-import qualified Data.Set as Set
-import           Data.Word (Word16)
+import           Data.Word ( Word16 )
 import           Network.MQTT.Packet
 import           Network.MQTT.Utils
+import qualified STMContainers.Set as TSet
+import qualified Data.Set as Set
```
Comment (rasendubi) [inline src/Network/MQTT/Client/Core.hs]: Should be unused now.

File: src/Network/MQTT/Client/Core.hs [hunk 23]
```diff
@@ -7,45 +7,53 @@ module Network.MQTT.Client.Core
...
+  { csSendQueue                   :: TQueue (Maybe Packet)
+  , csConnected                   :: TVar Bool
+  , csUsedPacketIdentifiers       :: TSet.Set PacketIdentifier
+  , csRandomVals                  :: TVar [Word16]
+  , csUnAckSentPublishPackets     :: TVar (Seq.Seq PublishPacket)
+  , csUnAckSentPubrelPackets      :: TVar (Seq.Seq PubrelPacket)
+  , csUnAckSentSubscribePackets   :: TVar (Seq.Seq SubscribePacket)
+  , csUnAckSentUnsubscribePackets :: TVar (Seq.Seq UnsubscribePacket)
+  , csUnAckReceivedPublishIds     :: TSet.Set PacketIdentifier
+  , csPublishCallback             :: TVar (Maybe (Topic       -> ByteString -> IO ()))
```
Comment (rasendubi) [inline src/Network/MQTT/Client/Core.hs]: Needless alignment.

File: src/Network/MQTT/Client/Core.hs [hunk 24]
```diff
@@ -89,138 +101,124 @@ runClient p is os = do
...
+            QoS0 ->
+              runPublishCallback state publishPacket
+            QoS1 -> do
+              runPublishCallback state publishPacket
+              atomically $ sendPacket state $ PUBACK (PubackPacket $ fromJust packetIdentifier)
+            QoS2 -> do
+              let packetIdentifier' = fromJust packetIdentifier
+              contains <- atomically $ TSet.lookup packetIdentifier' (csUnAckReceivedPublishIds state)
+              unless contains (runPublishCallback state publishPacket)
+              atomically $ do
```
Comment (rasendubi) [inline src/Network/MQTT/Client/Core.hs]: Merge two `atomically` in QoS 2 handling into one.

File: src/Network/MQTT/Client/Core.hs [hunk 25]
```diff
@@ -89,138 +101,124 @@ runClient p is os = do
...
+                TSet.insert packetIdentifier' $ csUnAckReceivedPublishIds state
```
Comment (rasendubi) [inline src/Network/MQTT/Client/Core.hs]: Don't insert if lookup succeeded.

File: src/Network/MQTT/Client/Core.hs [hunk 26]
```diff
@@ -89,138 +101,124 @@ runClient p is os = do
...
+                sendPacket state $ PUBREC (PubrecPacket packetIdentifier')
```
Comment (rasendubi) [inline src/Network/MQTT/Client/Core.hs]: I would do something like that
```haskell
present <- atomically $ do
  sendPacket state $ PUBREC (PubrecPacket packetIdentifier')

  let s = csUnAckReceivedPublishIds state
  present <- TSet.lookup packetIdentifier' s
  unless present $ TSet.insert packetIdentifier' s
  return present
unless present $ runPublishCallback state publishPacket
```

File: src/Network/MQTT/Client/Core.hs [hunk 27]
```diff
@@ -114,20 +107,23 @@ runPublishCallback state publishPacket = do
...
       (messageMessage $ publishMessage publishPacket)
 
-receiver :: InputStream Packet -> ClientState -> IO ()
-receiver is state = S.makeOutputStream handler >>= S.connect is
+authenticator :: ClientState -> InputStream Packet -> IO ()
+authenticator state is = do
+  -- Possible pattern-match failure is intended
+  Just (CONNACK ConnackPacket{..}) <- S.read is
+  when (connackReturnCode /= Accepted) $ throwIO MQTTError
+  receiver state is
```
Comment (rasendubi) [inline src/Network/MQTT/Client/Core.hs]:
```haskell
authenticator :: ClientState -> InputStream Packet -> IO ()
authenticator state is = do
  Just (CONNACK ConnackPacket{..}) <- S.read is
  unless (connackReturnCode /= Accepted) $ receiver state is
```

**Pull request merged**